src/core/model_abstract.jl

```diff
@@ -0,0 +1,251 @@
+- `rng = Random.default_rng()`: the random number generation stored and used by the model
+  in all calls to random functions. Accepts any subtype of `AbstractRNG`.
+- `warn=true`: some type tests for `AgentType` are done, and by default
+  warnings are thrown when appropriate.
+
+## Interface of `AgentBasedModel`
+
+Here we the most important information on how to query an instance of `AgentBasedModel`:
+
+- `model[id]` gives the agent with given `id`.
+- `abmproperties(model)` gives the `properies` container stored in the model.
+- `model.property`:  If the model properties is a dictionary with
+  key type `Symbol`, or if it is a composite type (`struct`), then the syntax
+  `model.property` will return the model property with key `:property`.
+- `abmrng(model)` will return the random number generator of the model.
+  It is strongly recommended to use `abmrng(model)` to all calls to `rand` and similar
+  functions, so that reproducibility can be established in your modelling workflow.
+- `abmscheduler(model)` will return the default scheduler of the model.
+
+Many more functions exist in the API page, such as [`allagents`](@ref).
+"""
+abstract type AgentBasedModel{S<:SpaceType, A<:AbstractAgent} end
+const ABM = AgentBasedModel
+
+function notimplemented(model)
+    error("Function not implemented for model of type $(nameof(typeof(model))) "*
+    "with space type $(nameof(typeof(abmspace(model))))")
+end
+
+###########################################################################################
+# %% Public methods. Must be implemented and are exported.
+###########################################################################################
+"""
+    model[id]
+    getindex(model::ABM, id::Integer)
+
+Return an agent given its ID.
+"""
+Base.getindex(m::ABM, id::Integer) = agent_container(m)[id]
+
+"""
+    abmrng(model::ABM)
+Return the random number generator stored in the `model`.
+"""
+abmrng(model::ABM) = getfield(model, :rng)
+
+"""
+    abmproperties(model::ABM)
+Return the properties container stored in the `model`.
+"""
+abmproperties(model::ABM) = getfield(model, :properties)
+
+"""
+    abmscheduler(model)
+Return the default scheduler stored in `model`.
+"""
+abmscheduler(model::ABM) = getfield(model, :scheduler)
+
+"""
+    allids(model)
+Return an iterator over all agent IDs of the model.
+"""
+allids(model) = eachindex(agent_container(model))
+
+"""
+    allagents(model)
+Return an iterator over all agents of the model.
+"""
+allagents(model) = values(agent_container(model))
+
+"""
+    nagents(model::ABM)
+Return the number of agents in the `model`.
+"""
+nagents(model::ABM) = length(allids(model))
+
+"""
+    nextid(model::ABM) → id
+Return a valid `id` for creating a new agent with it.
+"""
+nextid(model::ABM) = notimplemented(model)
+
+"""
+    random_agent(model) → agent
+Return a random agent from the model.
+"""
+random_agent(model) = model[rand(abmrng(model), allids(model))]
```

Comment:
As far as I know `eachindex` doesn't allocate. When I was implementing `allids` I chose to use `eachindex` over `keys` as `keys` allocates an "efficient" vector of indices when used on a vector i.e. one allocation but `eachindex` doesn't allocate at all and it was marginally faster because of this.

Reply:
```
julia> using BenchmarkTools

julia> model
UnkillableABM with 320 agents of type SchellingAgent
 space: GridSpaceSingle with size (20, 20), metric=chebyshev, periodic=false
 scheduler: Agents.Schedulers.Randomly
 properties: min_to_be_happy

julia> @btime random_agent($model)
  6.500 ns (0 allocations: 0 bytes)
SchellingAgent(36, (6, 19), false, 1)
```